flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,279 @@
+import org.apache.flink.util.function.TriConsumer;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Tests for the partition-lifecycle logic in the {@link TaskExecutor}.
+ */
+public class TaskExecutorPartitionLifecycleTest extends TestLogger {
+
+	private static final Time timeout = Time.seconds(10L);
+
+	@Rule
+	public final TemporaryFolder tmp = new TemporaryFolder();
+
+	private final JobID jobId = new JobID();
+	private TestingHighAvailabilityServices haServices;
+	private SettableLeaderRetrievalService jobManagerLeaderRetriever;
+	private TestingRpcService rpc;
+
+	@Before
+	public void setup() {
+		haServices = new TestingHighAvailabilityServices();
+		jobManagerLeaderRetriever = new SettableLeaderRetrievalService();
+		haServices.setResourceManagerLeaderRetriever(new SettableLeaderRetrievalService());
+		haServices.setJobMasterLeaderRetriever(jobId, jobManagerLeaderRetriever);
+		rpc = new TestingRpcService();
+	}
+
+	@Test
+	public void testPartitionReleaseAfterDisconnect() throws Exception {
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> taskExecutorGateway.disconnectJobManager(jobId, new Exception("test")),
+			(jobId, jobAwareShuffleEnvironment) -> assertFalse(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId)));
+	}
+
+	@Test
+	public void testPartitionReleaseAfterReleaseCall() throws Exception {
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> taskExecutorGateway.releasePartitions(jobId, Collections.singletonList(partitionId)),
+			(jobId, jobAwareShuffleEnvironment) -> assertFalse(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId)));
+	}
+
+	@Test
+	public void testPartitionReleaseAfterShutdown() throws Exception {
+		// don't do any explicit release action, so that the partition must be cleaned up on shutdown
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> { },
+			(jobId, jobAwareShuffleEnvironment) -> { });
+	}
+
+	private void testPartitionRelease(
+		TriConsumer<JobID, ResultPartitionID, TaskExecutorGateway> releaseAction,
+		BiConsumer<JobID, JobAwareShuffleEnvironmentImpl> postReleaseActionAssertion) throws Exception {
+
+		final TaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+
+		final TaskSlotTable taskSlotTable = new TaskSlotTable(
+			Collections.singletonList(ResourceProfile.UNKNOWN),
+			new TimerService<>(TestingUtils.defaultExecutor(), timeout.toMilliseconds()));
+
+		final JobLeaderService jobLeaderService = new JobLeaderService(taskManagerLocation, RetryingRegistrationConfiguration.defaultConfiguration());
```

Comment:
remove this to use default `JobLeaderService` in `TaskManagerServices `

Reply:
since it is also passed tot he JobLeaderService I'd like to keep it.